index.coffee

```diff
@@ -230,20 +225,46 @@ exports.DockerProgress = class DockerProgress
 						if not remoteId?
 							remoteId = remoteLayerIds[_.size(layerIds)]
 						layerIds[shortId] = remoteId
+
 					if status is 'Downloading'
-						currentDownloadedSize[shortId] = evt.progressDetail.current
-					else if status is 'Download complete' or status is 'Already exists'
+						downloadSize[shortId] = evt.progressDetail.current
+					else if status is 'Extracting'
+						extractSize[shortId] = evt.progressDetail.current
+					else if status is 'Download complete'
 						remoteId = layerIds[shortId]
-						completedSize += layerSizes[remoteId]
-						currentDownloadedSize[shortId] = 0
-					else if status.match(/^Status: Image is up to date for /)
-						completedSize = totalSize
-						currentDownloadedSize = {}
-
-					downloadedSize = completedSize + _.sum(currentDownloadedSize)
-					percentage = calculatePercentage(downloadedSize, totalSize)
-
-					onProgress(_.merge(evt, { downloadedSize, totalSize, percentage }))
+						downloadSize[shortId] = layerSizes[remoteId]
+					else if status is 'Pull complete'
+						remoteId = layerIds[shortId]
+						extractSize[shortId] = layerSizes[remoteId]
+					else if status is 'Already exists'
+						remoteId = layerIds[shortId]
+						downloadSize[shortId] = layerSizes[remoteId]
+						extractSize[shortId] = layerSizes[remoteId]
+
+					if status.match(/^Status: Image is up to date for /)
+						downloadedTotal = totalSize
+						extractedTotal = totalSize
+					else
+						downloadedTotal = _.sum(_.values(downloadSize))
+						extractedTotal = _.sum(_.values(extractSize))
+
+					total = totalSize * 2
+					completedSize = downloadedTotal + extractedTotal
+					downloadedPercentage = calculatePercentage(downloadedTotal, totalSize)
+					extractedPercentage = calculatePercentage(extractedTotal, totalSize)
+					percentage = calculatePercentage(completedSize, total)
+
+					onProgress _.merge evt, {
+						percentage
+						downloadedPercentage
+						extractedPercentage
+						downloadedSize: downloadedTotal
```

Comment:
I think it'd make sense for these to both be named the same - after all it's just exposing the data externally so it makes sense to have one consistent name (same for `extractedSize`)